File: helpers/clipboard_image.py
#!/usr/bin/env python3
"""Helper script to get image from macOS clipboard.

Outputs:
  - "image/png\n<base64_data>" if image found
  - "no_image" if no image
  - "error:<message>" on error
"""
import sys
import base64

try:
    from AppKit import NSPasteboard, NSPasteboardTypePNG, NSPasteboardTypeTIFF, NSBitmapImageRep, NSPNGFileType

    pasteboard = NSPasteboard.generalPasteboard()
    types = pasteboard.types()

    # Try PNG first
    if NSPasteboardTypePNG in types:
        data = pasteboard.dataForType_(NSPasteboardTypePNG)
        if data:
            print("image/png")
            print(base64.b64encode(bytes(data)).decode())
            sys.exit(0)

    # Try TIFF and convert to PNG
    if NSPasteboardTypeTIFF in types:
        data = pasteboard.dataForType_(NSPasteboardTypeTIFF)
        if data:
            bitmap = NSBitmapImageRep.imageRepWithData_(data)
            if bitmap:
                png_data = bitmap.representationUsingType_properties_(NSPNGFileType, None)
                if png_data:
                    print("image/png")
                    print(base64.b64encode(bytes(png_data)).decode())
                    sys.exit(0)

    print("no_image")

except Exception as e:
    print(f"error:{e}")
    sys.exit(1)

File: helpers/clipboard_image.js
// JXA script to get clipboard image
ObjC.import('AppKit');
ObjC.import('Foundation');

var pb = $.NSPasteboard.generalPasteboard;
var types = pb.types;

function output(str) {
    $.NSFileHandle.fileHandleWithStandardOutput.writeData(
        $.NSString.alloc.initWithString(str + '\n').dataUsingEncoding($.NSUTF8StringEncoding)
    );
}

// If clipboard has file URLs (e.g. copied from Finder), extract full paths.
// Finder copies include a thumbnail icon as public.tiff/public.png which we don't want.
if (types.containsObject('NSFilenamesPboardType')) {
    var plist = pb.propertyListForType('NSFilenamesPboardType');
    if (plist && plist.count > 0) {
        output('file_paths');
        for (var i = 0; i < plist.count; i++) {
            output(plist.objectAtIndex(i).js);
        }
        ObjC.import('stdlib');
        $.exit(0);
    }
}
if (types.containsObject('public.file-url')) {
    var urlStr = pb.stringForType('public.file-url');
    if (urlStr) {
        var url = $.NSURL.URLWithString(urlStr);
        if (url && url.isFileURL) {
            output('file_paths');
            output(url.path.js);
            ObjC.import('stdlib');
            $.exit(0);
        }
    }
}

// Check for PNG
if (types.containsObject('public.png')) {
    var data = pb.dataForType('public.png');
    if (data && data.length > 0) {
        var base64 = data.base64EncodedStringWithOptions(0).js;
        output('image/png');
        output(base64);
        ObjC.import('stdlib');
        $.exit(0);
    }
}

// Check for TIFF and convert to PNG
if (types.containsObject('public.tiff')) {
    var data = pb.dataForType('public.tiff');
    if (data && data.length > 0) {
        var bitmap = $.NSBitmapImageRep.imageRepWithData(data);
        if (bitmap) {
            var pngData = bitmap.representationUsingTypeProperties($.NSBitmapImageFileTypePNG, $());
            if (pngData && pngData.length > 0) {
                var base64 = pngData.base64EncodedStringWithOptions(0).js;
                output('image/png');
                output(base64);
                ObjC.import('stdlib');
                $.exit(0);
            }
        }
    }
}

output('no_image');
